README
NAME
    Mozilla::Backup - Backup utility for Mozilla profiles

REQUIREMENTS
    The following non-core modules are required:

      Archive::Tar
      Archive::Zip
      Compress::Zlib
      Config::IniFiles
      File::Temp
      IO::Zlib
      Log::Dispatch
      Module::Pluggable
      Params::Smart
      Regexp::Assemble;
      Regexp::Common
      Return::Value
      Test::More

    The Archive::* and *::Zlib modules are used by their respective plugins.

INSTALLATION
    Installation can be done using the traditional Makefile.PL or the newer
    Build.PL methods.

    Using Makefile.PL:

      perl Makefile.PL
      make test
      make install

    (On Windows platforms you should use `nmake' instead.)

    Using Build.PL (if you have Module::Build installed):

      perl Build.PL
      perl Build test
      perl Build install

SYNOPSIS
      $moz = Mozilla::Backup->new();
      $file = $moz->backup_profile("firefox", "default");

DESCRIPTION
    This package provides a simple interface to back up and restore the
    profiles of Mozilla-related applications such as Firefox or Thunderbird.

    More details are available in the module documentation.

REVISION HISTORY
    The following changes have been made since the last release:

    0.06  Sat May  7 2005
            - replaced Params::Validate with Params::Smart in constructors
            - revised internals for handling errors
            - Params::Smart 0.05 required
            - Compress::Zlib and IO::Zlib are now explicitly required
                  (for use by Archive::Tar)
            - plugins verify status of archive before operations
            - option to show version and license info in mozback script
            - plugins have smarter error handling/return values
            - methods return failure codes rather than croaking
            - uses Return::Value
            - _find_profile_path is now (publicly) documented
            - added support for Beonex profile
            - updated POD for README generation
            - improved code to find profile on Windows 9x/NT 
            - added code to munge prefs.js after restore_profile
            - uses IO::File
            - uses Regexp::Common
            - updated default excludes
            - updated documentation
            - set perms in FileCopy
            - added example to FileCopy
            - changed tests

    Details can be found in the Changes file.

KNOWN ISSUES
    This module is a prototype. Use at your own risk!

    Not all of the profile types have been tested, and are implemented based
    on information gleaned from sources which may or may not be accurate.

    The current version of this module only copies files and does little
    manipulation of any files, except for the profiles.ini and prefs.js to
    update some pathnames. This means that information specific to a profile
    on a machine such as extensions and themes is kept as-is, which may not
    be a good thing if a profile is restored to a different location or
    machine, or even application.

    (By default cache files are excluded from backups; there may be problems
    if cache files are restored to incompatible applications or machines.)

SEE ALSO
    Mozilla web site at http://www.mozilla.org.

AUTHOR
    Robert Rothenberg <rrwo at cpan.org>

  Suggestions and Bug Reporting

    Feedback is always welcome. Please use the CPAN Request Tracker at
    http://rt.cpan.org to submit bug reports.

    There is now a SourceForge project for this module at
    http://mozilla-backup.sourceforge.net/.

LICENSE
    Copyright (c) 2005 Robert Rothenberg. All rights reserved. This program
    is free software; you can redistribute it and/or modify it under the
    same terms as Perl itself.



Changes
Revision history for Perl extension Mozilla::Backup.

0.06  Sat May  7 2005
	- replaced Params::Validate with Params::Smart in constructors
	- revised internals for handling errors
	- Params::Smart 0.05 required
	- Compress::Zlib and IO::Zlib are now explicitly required
          (for use by Archive::Tar)
	- plugins verify status of archive before operations
	- option to show version and license info in mozback script
	- plugins have smarter error handling/return values
	- methods return failure codes rather than croaking
	- uses Return::Value
	- _find_profile_path is now (publicly) documented
	- added support for Beonex profile
	- updated POD for README generation
	- improved code to find profile on Windows 9x/NT 
	- added code to munge prefs.js after restore_profile
	- uses IO::File
	- uses Regexp::Common
	- updated default excludes
	- updated documentation
	- set perms in FileCopy
	- added example to FileCopy
	- changed tests

0.05  Sun Apr 24 2005
	* moved various functions from *::Backup to *::ProfilesIni
	  (using AUTOLOAD for deprecated function calls)
	- added type method
	- added (disabled) prereq test
	- added file portability tests
	- fixed bug with adding non-default profiles
	- fixed inconsistency with create_profile and profile_path
	- first profile created in a new ini file is always default
	- added Mozilla::ProfilesIni
	- added missing modules to pod-coverage test
	- additional tests
	- misc optimizations
	- fixed issue with newlines missing from croaked log messages
	- exclusions effect restore_profile method as well
	- user-configurable file exclusions
	- uses Regexp::Assemble
	- extensions are no longer ignored by default
	- improved lock file detection
	- added compression option for Zip plugin
	- added ability to pass options to plugin
	- added Tar plugin
	- uses Archive::Tar
	- chmod 0700 for mkdir calls
	- Default and StartWithLastProfile attributes updated appropriately
	- additional documentation/code tweaks
	- added profile_is_default method
	- additional error checking for profiles
	- support Default attribute for profiles
	- corrected typo in FileCopy.pm POD

0.04  Mon Apr 18 2005
	- added and revised tests
	- updated mozback script
	- _load_profile sub separated from constructor
	- documentation for FileCopy plugin
	- fixed issue with relative paths in FileCopy plugin
	- fixed bug that wiped pseudo profiles between sessions
	- various code tweaks
	- fixed bug that ignored all but last profile in profile.ini
	- documentation for Zip plugin
	- renamed methods required for plugins
	- added FileCopy plugin
	- uses File::Copy
	- debug flag adds screen output to log dispatcher
	- added various parameter and sanity checks
	- added internal _validate_type and _validate_profile methods
	- test tweaks
	- uses Params::Smart
	- moved log messages to plugin
	- implemented restore_profile method
	- added debug flag
	- tests are now only on the pseudo profile
	- added documentation of internal methods
	- added pseudo option to create pseudo profile type (for testing)
	- File::Temp now required for pseudo-profiles
	- added internal methods to create new profiles
	- added profile_exists method
	- fixed bug with croak/carp errors [rt.cpan.org #12234]
	- changed to use plugins instead of callbacks (callbacks removed!)
	- uses Module::Pluggable
	- various code comments
	- recognizes Windows-native Moz profiles in Cygwin
	- updated README
	- updated documentation to show SourceForge project site

0.03  Sun Apr 10 2005
	- added mozback script prototype
	- added profile_types method
	- smarter code for finding where profiles are
	* profile_section renamed to profile_id
	- added callbacks for using different archive formats
	- added profile_exists method
	- backup_profile dies if profile_is_locked
	- added profile_is_locked method
	- validation of other methods
	- more tests
	- option to control whether backups use relative pathnames
        - defaults not to use relative pathnames when profile's
          IsRelative flag is not set
	- misc code comments/tweaks
        - added Netscape profile
	- added Camino profile
	- added K-Meleon profile
	- added Phoenix profile (though not likely to be used)

0.02  Sat Apr  9 2005
	- released
	- lots of tweaking

0.01  Fri Apr  8 2005
	- original version; created by h2xs 1.23 with options
		-X -n Mozilla::Backup -v 0.01 -b 5.6.0



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Mozilla::Backup',
          'VERSION_FROM' => 'lib/Mozilla/Backup.pm',
          'PREREQ_PM' => {
                           'Archive::Tar' => '0',
                           'Archive::Zip' => '0',
                           'Carp' => '0',
                           'Compress::Zlib' => '0',
                           'Config::IniFiles' => '0',
                           'File::Copy' => '0',
                           'File::Find' => '0',
                           'File::Spec' => '0',
                           'File::Temp' => '0',
                           'IO::File' => '0',
                           'IO::Zlib' => '0',
                           'Log::Dispatch' => '1.6',
                           'Module::Pluggable' => '0',
                           'Params::Smart' => '0.05',
                           'Regexp::Assemble' => '0',
                           'Regexp::Common' => '1.8',
                           'Return::Value' => '0',
                           'Test::More' => '0'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'bin/mozback'
                         ],
          'PL_FILES' => {}
        )
;


lib/Mozilla/ProfilesIni.pm
=head1 NAME

Mozilla::ProfilesIni - Manipulation of Mozilla F<profiles.ini> files

=begin readme

=head1 REQUIREMENTS

The following non-core modules are required:

  Config::IniFiles
  Log::Dispatch
  Params::Smart
  Return::Value

=end readme

=head1 SYNOPSIS

  $path = Mozilla::ProfilesIni::_find_profile_path(
    home => $ENV{HOME}, 
    type => "firefox"
  );

  $ini = Mozilla::ProfilesIni->( path => $path );


=head1 DESCRIPTION

This module provides routines for parsing and manipulating Mozilla
F<profiles.ini> files.

The following methods are implemented:

=cut

package Mozilla::ProfilesIni;

require 5.006;

use strict;
use warnings;

use Carp;
use Config::IniFiles;
use File::Find;
use File::Spec;
use Log::Dispatch;
use Params::Smart 0.04;
use Return::Value;

# $Revision: 1.14 $

our $VERSION = '0.02';

=over

=item new

  $ini = Mozilla::ProfilesIni->new( $path );

  $ini = Mozilla::ProfilesIni->new( path => $path, %options );

The following options are supported:

=over

=item path

Path to where the F<profiles.ini> file is (excluding the actual
F<profiles.ini> file).  This is a required option.

=item log

A L<Log::Dispatch> object for receiving log messages.

=item debug

Sets an internal debug flag. (Not implemented at this time.)

=item create

Create a new F<profile.ini> file in L</path>.

=back

=cut

sub new {
  my $class = shift || __PACKAGE__;

  my %args  = Params(
   {
     name      => "path",
     required  => 1,
     name_only => 1,
   },
   {
     name     => "log",
     default  => Log::Dispatch->new(),
     callback => sub {
       my ($self, $name, $log) = @_;
       croak "invalid log sink"
	 unless ((ref $log) && $log->isa("Log::Dispatch"));
       return $log;
     },
     name_only => 1,
     required  => 0,
   },
   {
     name      => "debug",
     default   => 0,
     name_only => 1,
   },
   {
     name      => "create",
     default   => 0,
     name_only => 1,
   },
  )->args(@_);
  
  my $self  = {
    profiles  => { },
  };

  local ($_);

  foreach (qw( log debug create )) {
    $self->{$_} = $args{$_};
  }

  bless $self, $class;

  if ($self->{create}) {
    my $r = $self->_create_profile_ini( path => $args{path}, ignore => 1 );
    croak $r unless ($r);
  }

  my $r = $self->_read_profile_ini( path => $args{path} );
  croak $r unless ($r);

  return $self;
}

=begin internal

=item _create_profile_ini

  $ini->_create_profile_ini( path => $path, ignore => $ignore );

  $ini->_create_profile_ini( $path );

Creates a new F<profiles.ini> file in C<$path>.

By default it will die if a profiles file already exists, unless the
ignore flag is specified explicitly, in which case it will return
without creating a profile (or complaining).

=end internal

=cut

sub _create_profile_ini {
  my $self = shift;
  my %args = Params(qw( path ?+ignore ))->args(@_);
  my $path = File::Spec->rel2abs($args{path});

  unless (-d $path) {
    return failure $self->_log( 
      "cannot access psuedo profile directory: $path" );
  }

  my $ini_file = File::Spec->catfile($path, "profiles.ini" );
  if (-e $ini_file) {
    if ($args{ignore}) {
      return success;
    } else {
      return failure $self->_log(
        "a profile exists already at $path" );
    }
  } else {
    my $cfg = Config::IniFiles->new();
    $cfg->AddSection("General");
    $cfg->newval("General", "StartWithLastProfile", "");

    unless ($cfg->WriteConfig( $ini_file )) {
      return failure $self->_log(
        "unable to create pseudo configuration" );
    }

    unless (-e $ini_file) {
      return failure $self->_log(
	"unexpected error in creating pseudo configuration" );
    }
  }
  return success;
}

=begin internal

=item _read_profile_ini

  $ini->_read_profile_ini( path => $path );

  $ini->_read_profile_ini( $path );

Parses the F<profile.ini> in C<$path>.

This is called automatically by L</new>.

=end internal

=cut

sub _read_profile_ini {
  my $self = shift;
  my %args = Params(qw( path ))->args(@_);
  my $path = File::Spec->rel2abs($args{path});

  local ($_);

  $self->{profiles} = { };

  if (my $ini_file = _catfile($path, "profiles.ini")) {
    $self->{ini_file} = $ini_file;
    my $cfg = Config::IniFiles->new( -file => $ini_file );
    if ($cfg) {
      my $start_with_last = $cfg->val("General","StartWithLastProfile","");

      my $name = "";
      my $i = 0;
      while (my $profile = $cfg->val("Profile$i","Path")) {

        my $profile_path = _catdir($path, $profile);
	$profile_path = _catdir($profile) unless ($profile_path);

	if ($profile_path) {
	  my $data = {
	    ProfileId => "Profile$i",
	    Path      => $profile_path,
	  };

          unless ($name = $cfg->val("Profile$i", "Name")) {
	    return failure $self->_log(
	      "No name is defined for Profile$i");
          }

          # In nsToolkitProfileService.cpp, flags are "1" or ""

	  foreach (qw( Name IsRelative Default )) {
	    $data->{$_} = $cfg->val("Profile$i",$_, "");
	  }

	  $self->{profiles}->{$name} = $data;

	} else {
          # Do we warn instead of exit if there's bad data in profiles.ini?
	  return failure $self->_log(
	    "Bad Path: $profile_path not a directory");
	}
	$i++;
      }
      if ($start_with_last && $name) {
        $self->{profiles}->{$name}->{Default} = "1";
      }
    } else {
      return failure $self->_log(
       "Bad INI file: $ini_file");
    }
  } else {
    return failure $self->_log(
      "Cannot find profiles.ini in $path" );
  }
  return success;
}

=item create_profile

  $ini->create_profile( name => $name, is_default => $def, path => $path );

  $ini->create_profile( $name, $def, $path );

Creates a profile named C<$name> in C<$path>. If C<$path> is not
specified, it creates a relative profile in the F<Profiles>
subdirectory below the F<profiles.ini> file.

=cut

sub create_profile {
  my $self = shift;
  my %args = Params(qw( name ?is_default ?path ))->args(@_);
  my $name = $args{name};
  my $def  = $args{is_default} ? "1" : "";
  my $path = $args{path};

  local ($_);

  unless ($def || (keys %{$self->{profiles}})) {
    $def = "1";
    $self->_log( level => "info",
      message => "the only profile must be default" );
  }


  my $ini_file = $self->ini_file();
  if (-r $ini_file) {
    my @dirs = File::Spec->splitdir($ini_file);
    my $prof = File::Spec->catdir( @dirs[0..$#dirs-1], "Profiles" );
    unless (-d $prof) {
      $self->_log( level => "info",
        message => "creating directory $prof\n" );
      unless (mkdir $prof) {
	return failure $self->_log(
          "unable to create directory $prof" );      
      }

      # TODO - option whether to set perms; also a portable chmod

      chmod 0700, $prof;
    }

    # create a unique name

    # Note: Mozilla-style is to use "Profiles/$name/$random.slt" rather
    # than "Profiles/$random.$name"

    my $dir;
    unless ($path) {
      do {
        $dir = "";
        for (1..8) { $dir .= ("a".."z","0".."9")[int(rand(36))]; }
        $dir .= "." . $name;
        $path = File::Spec->catdir($prof, $dir);
      } while (-d $path);
    }

    $self->_log( level => "info",
        message => "creating directory $path\n" );
    unless (mkdir $path) {
	return failure $self->_log(
          "unable to create directory $path" );
    }
    chmod 0700, $path;

    # BUG/TODO - We need to check how Mozilla etc. handles profile ids

    my $id = "Profile" . scalar( keys %{$self->{profiles}} );

    foreach (keys %{$self->{profiles}}) {
      if ($self->{profiles}->{$_}->{ProfileId} eq $id) {
        return failure $self->_log(
          "Profile Id conflict" );
      }
    }

    my $cfg = Config::IniFiles->new( -file => $ini_file );

    # update profile default flags

    foreach (keys %{$self->{profiles}}) {
      my $data = $self->{profiles}->{$_};
      $data->{Default} = "", if ($def);
      if (defined $cfg->val($data->{ProfileId}, "Default")) {
	$cfg->setval($data->{ProfileId}, "Default", $data->{Default});
      } else {
	$cfg->newval($data->{ProfileId}, "Default", $data->{Default});
      }
    }

    if ($def) {
      $cfg->setval("General", "StartWithLastProfile", "1");
    }
    else {
      $cfg->setval("General", "StartWithLastProfile", "");
    }

    my $data = {
      ProfileId  => $id,
      Name       => $name,
      IsRelative => (($dir) ? "1" : ""),
      Default    => $def,
      Path       => (($dir) ? ("Profiles/" . $dir) : $path),
    };

    $cfg->AddSection($id);
    foreach (qw( Name IsRelative Path Default )) {
      $cfg->newval($id, $_, $data->{$_});
    }

    $data->{Path} = $path;
    $self->{profiles}->{$name} = $data;

    # TODO/BUG? - Make sure IsRelative paths are not changed to
    # absolute paths when rewritten!

    unless ($cfg->RewriteConfig) {
      return failure $self->_log(
        "Unable to update INI file" );
    }
  }
  else {
    return failure $self->_log(
      "cannot find INI file $ini_file" );
  }
  return success;
}

=item ini_file

   $path = $ini->ini_file();

Returns the path to the F<profiles.ini> file.

=cut

sub ini_file {
  my $self = shift;
  return $self->{ini_file};
}

=item profile_names

  @names = $ini->profile_names($type);

Returns the names of profiles associated with the type.

=cut

sub profile_names {
  my $self = shift;
  return (keys %{$self->{profiles}});
}

=item profile_exists

  if ($ini->profile_exists($name)) { ... }

Returns true if a profile exists.

=cut

sub profile_exists {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};
  return (exists $self->{profiles}->{$name});
}

=item profile_is_relative

  if ($ini->profile_is_relative($name)) { ... }

Returns the "IsRelative" flag for the profile.

=cut

sub profile_is_relative {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};
  # TODO - validate profile name
  return $self->{profiles}->{$name}->{IsRelative};
}

=item profile_path

  $path = $ini->profile_path($name);

Returns the pathname of the profile.

=cut

sub profile_path {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};

  # TODO - validate profile name

  my $path = $self->{profiles}->{$name}->{Path};
#   if ($self->profile_is_relative($name) && (!-d $path)) {
#     my @dirs = File::Spec->splitdir( $self->ini_file );
#     $path = File::Spec->catdir( @dirs[0..$#dirs-1], $path );
#   }
  return $path;
}

=item profile_is_default

  if ($ini->profile_is_default($name)) { ... }

Returns the "Default" flag for the profile.

=cut

sub profile_is_default {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};
  return $self->{profiles}->{$name}->{Default};
}

=item profile_id

  $section = $ini->profile_id($name);

Returns the L</ini_file> identifier of the profile.

=cut

sub profile_id {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};
  return $self->{profiles}->{$name}->{ProfileId};
}

=item profile_is_locked

  if ($ini->profile_is_locked($name)) { ... }

Returns true if there is a lock file in the profile.

=cut

sub profile_is_locked {
  my $self = shift;
  my %args = Params(qw( name ))->args(@_);
  my $name = $args{name};
  foreach ('parent.lock', 'lock', '.parentlock') {
    if (_catfile($self->profile_path(name => $name), $_ )) {
      return 1;
    }      
  }
  return;
}

=begin internal

=item _catdir

  $path = _catdir( @names );

Returns the C<$path> if the concatenation of C<@names> exists as a
directory, or C<undef> otherwise.

=item _catfile

  $path = _catdir( @names );

Returns the C<$path> if the concatenation of C<@names> exists as a
file, or C<undef> otherwise.

=end internal

=cut

sub _catdir {
  if ($_[0]) { # otherwise blank "" is translated to root directory
    my $path = File::Spec->catdir(@_);
    return (-d $path) ? $path : undef;
  }
  else {
    return;
  }
}

sub _catfile {
  my $path = File::Spec->catfile(@_);
  return (-r $path) ? $path : undef;
}

=item _find_profile_path

  $path = _find_profile_path( home => $home, type => $type );

  $path = _find_profile_path( $home, $type );

Looks for a directory corresponding to where profile type of C<$type>
should be, generally somewhere in the C<$home> directory, where
C<$home> is the platform-specific "home" directory (not necessarily
C<$ENV{HOME}>).

Returns C<undef> if no path for that type was found.

In cases where profile paths cannot be found, use the C<MOZILLA_HOME>
or C<appname_HOME> environment variable to indicate where it is.

=cut

sub _find_profile_path {
  my %args = Params(qw( home type ))->args(@_);
  my $home = $args{home};
  my $type = $args{type};

  # Known Issue: the first profile that it finds for a type is the one
  # it uses. If for some reason there are profiles for the same
  # application in multiple places (maybe due to an upgrade), it will
  # use the first that it finds.

  my $path;

  # The MOZILLA_HOME environment variables are for OS/2, but putting
  # them here first allows one to override settings.

  # Problem? On some OSs, these may indicate where Mozilla binaries are,
  # and not profiles!

  if ($path = _catdir($ENV{uc($type)."_HOME"})) {
    return $path;
  }
  if ($path = _catdir($ENV{MOZILLA_HOME}, ucfirst($type))) {
    return $path;
  }
#   if ($path = _catdir($ENV{MOZILLA_FIVE_HOME})) {
#     return $path;
#   }
#   if ($path = _catdir($ENV{MOZILLA_FIVE_HOME}, ucfirst($type))) {
#     return $path;
#   }

  if ($path = _catdir($home, "\.$type")) {
    return $path;
  }
  if ($path = _catdir($home, "\.mozilla", $type)) {
    return $path;
  }
  if ($path = _catdir($home, ucfirst($type))) {
    return $path;
  }
  if ($path = _catdir($home, "Mozilla", ucfirst($type))) {
    return $path;
  }

  if ($^O eq "darwin") {
    if ($path = _catdir($home, "Library", "Application Support",
			ucfirst($type))) {
      return $path;
    }
    if ($path = _catdir($home, "Library", "Application Support",
			"Mozilla", ucfirst($type))) {
      return $path;
    }
    if ($path = _catdir($home, "Library", ucfirst($type))) {
      return $path;
    }
    if ($path = _catdir($home, "Library", "Mozilla", ucfirst($type))) {
      return $path;
    }
  }
  elsif ($^O eq "MSWin32") {
    my $program_files = $ENV{ProgramFiles} || "Program Files";
    if ($path = _catdir($program_files, ucfirst($type))) {
      return $path;
    }
    if ($path = _catdir($program_files, "Mozilla", ucfirst($type))) {
      return $path;
    }
  }

  # If we're here in Cygwin, it means that Mozilla builds are probably
  # native-Windows instead of Cygwin. So we need to look in the
  # Windows native drive.

  # Known Issue: if you have separate Cygwin and Windows Moz profiles,
  # then it will recognize the Cygwin profile first.

  elsif ($^O eq "cygwin") {
    if ((caller(1))[3] !~ /_find_profile_path$/) {
      $home = $ENV{APPDATA}; # Win 2000/XP/2003
      $home =~ s/^(\w):/\/cygdrive\/$1/;
      return _find_profile_path($home,$type);
    }
  }

  return;
}

=begin internal

=item _log

  $moz->_log( $message, $level );

  $moz->_log( $message => $message, level => $level );

Logs an event to the dispatcher. If C<$level> is unspecified, "error"
is assumed.

=end internal

=cut

sub _log {
  my $self = shift;
  my %args = Params(qw( message ?level=error ))->args(@_);
  my $msg  = $args{message};

  # we want log messages to always have a newline, but not necessarily
  # the returned value that we pass to carp/croak/return value

  $args{message} .= "\n" unless ($args{message} =~ /\n$/);
  $self->{log}->log(%args) if ($self->{log});
  return $msg;    # when used by carp/croak/return value
}

=back

=head1 CAVEATS

This module is a prototype. Use at your own risk!

=head1 SEE ALSO

L<Mozilla::Backup>

=head1 AUTHOR

Robert Rothenberg <rrwo at cpan.org>

=head1 LICENSE

Copyright (c) 2005 Robert Rothenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut

1;

__END__


Build.PL
# $Revision: 1.20 $

use Module::Build;

my $build = Module::Build->new
    (
     module_name => 'Mozilla::Backup',
     license => 'perl',
     requires => {
       'perl'             => '5.6.0',
       'Carp'             => 0,
       'Archive::Tar'     => 0,
       'Archive::Zip'     => 0,
       'Compress::Zlib'   => 0,         # for Archive::Zip and IO::Zlib 
       'Config::IniFiles' => 0,
       'File::Copy'       => 0,
       'File::Find'       => 0,
       'File::Spec'       => 0,
       'File::Temp'       => 0,
       'IO::File'         => 0,
       'IO::Zlib'         => 0,         # for Archive::Tar
       'Log::Dispatch'    => 1.6,
       'Module::Pluggable' => 0,
       'Params::Smart'    => 0.05,
       'Regexp::Assemble' => 0,
       'Regexp::Common'   => 1.8,
       'Return::Value'    => 0,
     },
     recommends => {
       'Getopt::Long'          => 0,    # for script
       'Log::Dispatch::Screen' => 0,    # for debugging
       'Pod::Readme'           => 0.02, # for generating distro
       'Test::Pod::Coverage'   => 0,    # for tests
       'Test::Pod'             => 1.00,
       'Test::Portability::Files' => 0,
#        'Test::Prereq'   => 0,
     },
     build_requires => {
       'Test::More'     => 0,
     },
     script_files       => [
       'bin/mozback',
     ],
     create_readme      => 0,
     create_makefile_pl => 'traditional',
     dist_author        => 'Robert Rothenberg <rrwo at cpan.org>',
     sign               => 1,
);

$build->create_build_script;

